sw/device/tests/dif/dif_gpio_test.cc

```diff
@@ -0,0 +1,420 @@
+  EXPECT_WRITE32(GPIO_INTR_CTRL_EN_LVLLOW_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_CTRL_EN_INPUT_FILTER_REG_OFFSET, 0);
+  EXPECT_WRITE32(GPIO_INTR_STATE_REG_OFFSET, 0xFFFFFFFFu);
+
+  EXPECT_EQ(dif_gpio_reset(&gpio_), kDifGpioResultOK);
+}
+
+// Read tests
+class ReadTest : public DifGpioTestWithDevice {};
+
+TEST_F(ReadTest, NullArgs) {
+  uint32_t out_arg_uint32_t;
+  bool out_arg_bool;
+
+  EXPECT_EQ(dif_gpio_all_read(nullptr, &out_arg_uint32_t),
+            kDifGpioResultInvalidArgument);
+  EXPECT_EQ(dif_gpio_all_read(&gpio_, nullptr), kDifGpioResultInvalidArgument);
+  EXPECT_EQ(dif_gpio_all_read(nullptr, nullptr), kDifGpioResultInvalidArgument);
+
+  EXPECT_EQ(dif_gpio_pin_read(nullptr, 0, &out_arg_bool),
+            kDifGpioResultInvalidArgument);
+  EXPECT_EQ(dif_gpio_pin_read(&gpio_, 0, nullptr),
+            kDifGpioResultInvalidArgument);
+  EXPECT_EQ(dif_gpio_pin_read(nullptr, 0, nullptr),
+            kDifGpioResultInvalidArgument);
+}
+
+TEST_F(ReadTest, AllPins) {
+  const uint32_t exp_pin_values = 0xA5A5A5A5;
+  EXPECT_READ32(GPIO_DATA_IN_REG_OFFSET, exp_pin_values);
+
+  uint32_t pin_values = 0;
+  EXPECT_EQ(dif_gpio_all_read(&gpio_, &pin_values), kDifGpioResultOK);
+  EXPECT_EQ(pin_values, exp_pin_values);
+}
+
+TEST_F(ReadTest, SinglePin) {
+  for (uint32_t pin = 0; pin < 32; ++pin) {
+    for (const bool exp_pin_val : {true, false}) {
+      uint32_t const reg_val =
+          exp_pin_val ? AllZerosExcept(pin) : AllOnesExcept(pin);
+      EXPECT_READ32(GPIO_DATA_IN_REG_OFFSET, reg_val);
+      bool pin_val = !exp_pin_val;
+      EXPECT_EQ(dif_gpio_pin_read(&gpio_, pin, &pin_val), kDifGpioResultOK);
+      EXPECT_EQ(pin_val, exp_pin_val);
+    }
+  }
+}
+
+// Write tests
+class WriteTest : public DifGpioTestWithDevice {};
+
+TEST_F(WriteTest, NullArgs) {
+  EXPECT_EQ(dif_gpio_all_write(nullptr, 0xFFFFFFFF),
+            kDifGpioResultInvalidArgument);
+  EXPECT_EQ(dif_gpio_pin_write(nullptr, 0, true),
+            kDifGpioResultInvalidArgument);
+  EXPECT_EQ(dif_gpio_masked_write(nullptr, 0xFFFFFFFF, 0xFFFFFFFF),
+            kDifGpioResultInvalidArgument);
+}
+
+TEST_F(WriteTest, AllPins) {
+  uint32_t const val = 0xA5A5A5A5;
+  EXPECT_WRITE32(GPIO_DIRECT_OUT_REG_OFFSET, val);
+
+  EXPECT_EQ(dif_gpio_all_write(&gpio_, val), kDifGpioResultOK);
+}
+
+TEST_F(WriteTest, SinglePin) {
+  EXPECT_WRITE32(GPIO_MASKED_OUT_LOWER_REG_OFFSET, {{16, 1}, {0, 1}});
+  EXPECT_EQ(dif_gpio_pin_write(&gpio_, 0, true), kDifGpioResultOK);
+
+  EXPECT_WRITE32(GPIO_MASKED_OUT_LOWER_REG_OFFSET, {{31, 1}, {15, 0}});
+  EXPECT_EQ(dif_gpio_pin_write(&gpio_, 15, false), kDifGpioResultOK);
+
+  EXPECT_WRITE32(GPIO_MASKED_OUT_UPPER_REG_OFFSET, {{16, 1}, {0, 1}});
+  EXPECT_EQ(dif_gpio_pin_write(&gpio_, 16, true), kDifGpioResultOK);
+
+  EXPECT_WRITE32(GPIO_MASKED_OUT_UPPER_REG_OFFSET, {{31, 1}, {15, 0}});
+  EXPECT_EQ(dif_gpio_pin_write(&gpio_, 31, false), kDifGpioResultOK);
+}
+
+TEST_F(WriteTest, Masked) {
```

Comment:
Could you please add comments of how `EXPECT_WRITE32` call values relate to `dif_gpio_masked_write` call values, as it is not obvious without reading the spec or the code of `dif_gpio` functions.

Reply:
If it's not clear from the spec, I think that should be raised with the HW folks in an issue, too.

Reply:
I think the spec and the DIF library is quite clear. I added a comment that explains this relationship to make the tests easier to read.

Reply:
Somehow I got confused by this test the first time around, now having a second look - it is pretty clear what is happening. Maybe `0xAAAABBBB` values would be more demonstrative in this case (feel free to ignore, if you think that current values are better)?